src/Language/DifferentialDatalog/TypeInference.hs

```diff
@@ -0,0 +1,637 @@
+
+--- 'if (e1){e2}else{e3}'
+--
+-- '|e1|=Bool and |e2|=|e3|'
+exprConstraints_ de@(DDExpr ctx (E e@EITE{..})) =
+    [ TETypeOfExpr de === TEBool
+    , TETypeOfExpr (DDExpr (CtxITEThen e ctx) exprThen) === TETypeOfExpr (DDExpr (CtxITEElse e ctx) exprElse) ]
+
+-- 'for (v in e1) {e2}'
+--
+-- 'is_iterable |e1| and |var v|=iterator_type |e1| and |e| = ()'
+exprConstraints_ de@(DDExpr ctx (E e@EFor{..})) =
+    [ deIsIterable (DDExpr (CtxForIter e ctx) exprIter) (TETypeOfVar $ ForVar ctx e)
+    , TETypeOfExpr de === TETuple (Just 0) M.empty ]
+
+-- Assignment: 'e1=e2'.
+--
+-- '|e|=Tuple0 and |e1|=|e2|'.
+exprConstraints_ de@(DDExpr ctx (E e@ESet{..})) =
+    [ TETypeOfExpr de === TETuple (Just 0) M.empty
+    , TETypeOfExpr (DDExpr (CtxSetL e ctx) exprLVal) === TETypeOfExpr (DDExpr (CtxSetR e ctx) exprRVal) ]
+
+-- 'return e1'
+--
+-- '|e1|=retType'
+exprConstraints_ (DDExpr ctx (E e@EReturn{..})) =
+    [TETypeOfExpr (DDExpr (CtxReturn e ctx) exprRetVal) === typeToTExpr funcType]
+    where
+    Just Function{..} = ctxInFunc ctx
+
+-- Binary operator 'e1 op e2', where `op` is one of '==, !=, <, >, <=, >='.
+--
+-- '|e1| = |e2| ans |e| = Bool'
+exprConstraints_ de@(DDExpr ctx (E e@EBinOp{..})) | elem exprBOp [Eq, Neq, Lt, Lte, Gt, Gte] =
+    [ TETypeOfExpr l === TETypeOfExpr r
+    , TETypeOfExpr de === TEBool ]
+
+-- Binary operator 'e1 op e2', where `op` is one of '&&, ||, =>'.
+--
+-- '|e1| = |e2| = Bool and |e| == Bool'
+                                                | elem exprBOp [And, Or, Impl] =
+    [ TETypeOfExpr l === TETypeOfExpr (DDExpr (CtxBinOpR e ctx) exprRight)
+    , TETypeOfExpr l === TEBool
+    , TETypeOfExpr de === TEBool ]
+
+-- Binary operator 'e1 op e2', where `op` is one of '+, -, *. /'.
+--
+-- '|e| = |e1| = |e2| and is_num|e1|'
+                                                | elem exprBOp [Plus, Minus, Times, Div] =
+    [ TETypeOfExpr l === TETypeOfExpr r
+    , TETypeOfExpr de === TETypeOfExpr l
+    , deIsNum $ DDExpr (CtxBinOpL e ctx) exprLeft ]
+
+-- Binary operator 'e1 % e2'.
+--
+-- '|e1| = |e2| and is_int(|e1|) and |e| = |e1|'
+                                                | exprBOp == Mod =
+    [ TETypeOfExpr l === TETypeOfExpr r
+    , deIsInt l
+    , TETypeOfExpr de === TETypeOfExpr l ]
+
+-- Binary operator 'e1 op e2', where `op` is one of '<<, >>'.
+--
+-- 'is_int |e1| and is_int |e2| and |e| = |e1|'
+                                                | elem exprBOp [Plus, Minus] =
+    [ deIsInt l
+    , deIsInt r
+    , TETypeOfExpr de === TETypeOfExpr l ]
+
+-- Binary operator 'e1 op e2', where `op` is one of '|,&,^'.
+--
+-- '|e| = |e1| = |e2| and is_bits|e1|'
+                                                | elem exprBOp [BOr, BAnd, BXor] =
+    [ TETypeOfExpr l === TETypeOfExpr r
+    , TETypeOfExpr de === TETypeOfExpr l
+    , deIsBits l ]
+
+-- Binary operator 'e1 ++ e2'.
+--
+-- is_String(|e1|) and is_String(|e2|) or is_Bit |e1| and is_Bit |e2|
```

Comment:
I was expecting a constraint on the lengths for the bits, but if you go bottom up it's not necessary

Reply:
you're right, and this is what the implementation actually does. I will update the comment.